File: 0783-minimum-distance-between-bst-nodes/0783-minimum-distance-between-bst-nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        self.mindif=float('inf')
        self.prev=None
        self.inorder(root)
        return self.mindif
    def inorder(self,root: TreeNode):
        if root is None:
            return
        self.inorder(root.left)
        if self.prev is not None:
            self.mindif=min(self.mindif,root.val-self.prev)
        self.prev=root.val
        self.inorder(root.right)
      

File: 0783-minimum-distance-between-bst-nodes/0783-minimum-distance-between-bst-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void fun(TreeNode* root,int& preval,int& mindif){
        if (root==nullptr) return;
        fun(root->left,preval,mindif);
        if (preval != -1){
            mindif=min(mindif,root->val-preval);
        }
        preval=root->val;
        fun(root->right,preval,mindif);
    }
    
    int minDiffInBST(TreeNode* root) {
        int mindif=INT_MAX;
        int preval=-1;
        fun(root,preval,mindif);
        return mindif;
    }
};